cdrwtool/cdrwtool.c

```diff
@@ -73,80 +72,87 @@ void dump_sense(unsigned char *cdb, struct request_sense *sense)
 	}
 }
 
-int wait_cmd(int fd, struct cdrom_generic_command *cgc, unsigned char *buf,
-			 int dir, int timeout)
+int wait_cmd_sense(int fd, struct cdrom_generic_command *cgc,
+		   unsigned char *buf, int dir, int timeout,
+		   struct request_sense *sense)
 {
-	struct request_sense sense;
 	int ret;
 
 	if (timeout <= 0)
 		timeout = 500;
 
-	memset(&sense, 0, sizeof(sense));
+	memset(sense, 0, sizeof(*sense));
 
 	cgc->buffer = buf;
 	cgc->data_direction = dir;
-	cgc->sense = &sense;
+	cgc->sense = sense;
 	cgc->timeout = timeout;
 
 	ret = ioctl(fd, CDROM_SEND_PACKET, cgc);
-	if (ret)
+	if (ret && cgc->cmd[0] != GPCMD_TEST_UNIT_READY)
 	{
-		perror("wait_cmd");
+		perror("wait_cmd_sense");
 		dump_sense(cgc->cmd, cgc->sense);
 	}
 	return ret;
 }
 
-int sigfd;
-
-void sig_progress(int sig)
+int wait_cmd(int fd, struct cdrom_generic_command *cgc, unsigned char *buf,
+	     int dir, int timeout)
 {
-	struct cdrom_generic_command cgc;
 	struct request_sense sense;
-	static int did = 0;
 	int ret;
 
-	if (sig != SIGALRM)
-		return;
-
-	memset(&cgc, 0, sizeof(cgc));
-	memset(&sense, 0, sizeof(sense));
-
-	cgc.sense = &sense;
-	ret = wait_cmd(sigfd, &cgc, NULL, CGC_DATA_NONE, WAIT_PC);
-
-	if ((ret || !(sense.sks[0] & 0x80)) && !did) {
-		printf("Progress indicator not implemented on this drive\n");
-		printf("Don't access drive until operation has completed\n");
-		progress = 101;
-		return;
-	}
-
-	progress = ((sense.sks[1] << 8 | sense.sks[2]) * 100) / 0xffff;
-	did = 1;
-
-	printf("%02d%% complete\n", progress);
-	if (progress == 99) {
-		progress = 101;
-		return;
-	}
-	alarm(2);
+	ret = wait_cmd_sense(fd, cgc, buf, dir, timeout, &sense);
+	cgc->sense = NULL; /* invalidate pointer to local variable */
+	return ret;
 }
 
-void print_completion_info(int fd)
+static int wait_for_unit_ready(int fd, char *what, int timeout)
 {
-	sigfd = fd;
-	/* we can only poll sense for non-blocking commands */
-#if USE_IMMED == 0
-	return;
-#else
-	progress = 0;
-	signal(SIGALRM, sig_progress);
-	alarm(5);
-	while (progress < 100)
+	int ret, progress, reported = 0;
+	int interval_elapsed = 0, total_elapsed = 0, interval_length = 5;
+	unsigned char buffer[16];
+	struct cdrom_generic_command cgc;
+	struct request_sense sense;
+
+	memset(&cgc, 0, sizeof(cgc));
+	memset(buffer, 0, sizeof(buffer));
+
+	while (total_elapsed < timeout) {
```

Comment:
... and then just this place needs to be update to convert `timeout` from hz to `timeout_sec` in seconds.